File: src/cz/vutbr/fit/interlockSim/util/UnorientedGraph.java
/* Brno University of Technology
 * Faculty of Information Technology
 * 
 * BSc Thesis	2006/2007
 * 
 * Railway Interlocking Simulator
 * 
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.util;

import java.util.Collection;
import java.util.Map;
import java.util.Set;

/**
 * ADT Graph interface
 *
 * @param <N> nodes
 * @param <E> edges
 */
public interface UnorientedGraph<N,E>{
	
	/**
	 * Put element to ExtendedUnorientedGraph
	 * @param first
	 * @param second
	 * @param value
	 */
	public void put(N first, N second, E value);
	
	/**
	 * get edge from graph
	 * @param first
	 * @param second
	 * @return edge
	 */
	public E get(N first, N second);
	
	/**
	 * @param node
	 * @return all edges, which joins with node
	 */
	public Collection<E> get(N node);

	/**
	 * Remove edge from graph
	 * @param first
	 * @param second
	 * @return edge
	 */
	public E remove(N first, N second);

	/**
	 * Remove node and all edges, which join node with others
	 * @param node
	 * @return all edges which were removed
	 */
	public Collection<E> removeAll(N node);

	/**
	 * Remove edge from graph
	 * @param edge
	 * @return nodes which were joined with this edge  
	 */
	public Collection<N> remove(E edge);
	
	/**
	 * @return all nodes, which is in graph
	 */
	public Set<N> nodeSet();

	/**
	 * @return {@link Map#isEmpty()}
	 */
	public boolean isEmpty();
	
	/**
	 * @see Map#values()
	 * @return all edges in graph
	 */
	public Collection<E> values();

	/**
	 * @return number of edges
	 */
	public int size();

	/**
	 * Remove all edges from graph
	 */
	public void clear();
	
	/**
	 * @param node1 
	 * @param node2 
	 * @return true If graph contains edge between nodes
	 */
	public boolean contains(N node1, N node2);
	
	/**
	 * @param edge 
	 * @return true If graph contains edge
	 */
	public boolean containsEdge(E edge);
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/util/UnorientedGraph.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.util

import java.util.Collection
import java.util.Set

/**
 * ADT Graph interface
 *
 * @param <N> nodes
 * @param <E> edges
 */
interface UnorientedGraph<N, E> {
	/**
	 * Put element to ExtendedUnorientedGraph
	 * @param first
	 * @param second
	 * @param value
	 */
	fun put(
		first: N,
		second: N,
		value: E
	)

	/**
	 * get edge from graph
	 * @param first
	 * @param second
	 * @return edge
	 */
	operator fun get(
		first: N,
		second: N
	): E?

	/**
	 * @param node
	 * @return all edges, which joins with node
	 */
	fun get(node: N): Collection<E>

	/**
	 * Remove edge from graph
	 * @param first
	 * @param second
	 * @return edge
	 */
	fun remove(
		first: N,
		second: N
	): E?

	/**
	 * Remove node and all edges, which join node with others
	 * @param node
	 * @return all edges which were removed
	 */
	fun removeAll(node: N): Collection<E>

	/**
	 * Remove edge from graph
	 * @param edge
	 * @return nodes which were joined with this edge
	 */
	fun remove(edge: E): Collection<N>

	/**
	 * @return all nodes, which is in graph
	 */
	fun nodeSet(): Set<N>

	/**
	 * @return {@link Map#isEmpty()}
	 */
	fun isEmpty(): Boolean

	/**
	 * @see Map#values()
	 * @return all edges in graph
	 */
	fun values(): Collection<E>

	/**
	 * @return number of edges
	 */
	fun size(): Int

	/**
	 * Remove all edges from graph
	 */
	fun clear()

	/**
	 * @param node1
	 * @param node2
	 * @return true If graph contains edge between nodes
	 */
	fun contains(
		node1: N,
		node2: N
	): Boolean

	/**
	 * @param edge
	 * @return true If graph contains edge
	 */
	fun containsEdge(edge: E): Boolean
}
